lib/std/enums.zig

```diff
@@ -0,0 +1,1281 @@
+    }
+}
+
+test "std.enums.valuesFromFields" {
+    const E = extern enum { a, b, c, d = 0 };
+    const fields = valuesFromFields(E, &[_]EnumField{
+        .{ .name = "b", .value = undefined },
+        .{ .name = "a", .value = undefined },
+        .{ .name = "a", .value = undefined },
+        .{ .name = "d", .value = undefined },
+    });
+    testing.expectEqual(E.b, fields[0]);
+    testing.expectEqual(E.a, fields[1]);
+    testing.expectEqual(E.d, fields[2]); // a == d
+    testing.expectEqual(E.d, fields[3]);
+}
+
+/// Returns the set of all named values in the given enum, in
+/// declaration order.
+pub fn values(comptime E: type) []const E {
+    return comptime valuesFromFields(E, @typeInfo(E).Enum.fields);
+}
+
+test "std.enum.values" {
+    const E = extern enum { a, b, c, d = 0 };
+    testing.expectEqualSlices(E, &.{.a, .b, .c, .d}, values(E));
+}
+
+/// Returns the set of all unique named values in the given enum, in
+/// declaration order.  For repeated values in extern enums, only the
+/// first name for each value is included.
+pub fn uniqueValues(comptime E: type) []const E {
+    return comptime valuesFromFields(E, uniqueFields(E));
+}
+
+test "std.enum.uniqueValues" {
+    const E = extern enum { a, b, c, d = 0, e, f = 3 };
+    testing.expectEqualSlices(E, &.{.a, .b, .c, .f}, uniqueValues(E));
+
+    const F = enum { a, b, c };
+    testing.expectEqualSlices(F, &.{.a, .b, .c}, uniqueValues(F));
+}
+
+/// Returns the set of all unique field values in the given enum, in
+/// declaration order.  For repeated values in extern enums, only the
+/// first name for each value is included.
+pub fn uniqueFields(comptime E: type) []const EnumField {
+    comptime {
+        const info = @typeInfo(E).Enum;
+        const raw_fields = info.fields;
+        // Only extern enums can contain duplicates,
+        // so fast path other types.
+        if (info.layout != .Extern) {
+            return raw_fields;
+        }
+
+        var unique_fields: []const EnumField = &[_]EnumField{};
+    outer:
+        for (raw_fields) |candidate| {
+            for (unique_fields) |u| {
+                if (u.value == candidate.value)
+                    continue :outer;
+            }
+            unique_fields = unique_fields ++ &[_]EnumField{candidate};
+        }
+
+        return unique_fields;
+    }
+}
+
+/// Determines the length of a direct-mapped enum array, indexed by
+/// @intCast(usize, @enumToInt(enum_value)).  The enum must be exhaustive.
+/// If the enum contains any fields with values that cannot be represented
+/// by usize, a compile error is issued.  The max_unused_slots parameter limits
+/// the total number of items which have no matching enum key (holes in the enum
+/// numbering).  So for example, if an enum has values 1, 2, 5, and 6, max_unused_slots
+/// must be at least 3, to allow unused slots 0, 3, and 4.
+fn directEnumArrayLen(comptime E: type, comptime max_unused_slots: comptime_int) comptime_int {
```

Comment:
This should also be offset by a `min`

Reply:
The intent of this api is to be directly indexed by the integer value of the enum, hence the `direct` part of the name.  The `max_unused_slots` parameter ensures you don't have more padding than you intended.  If you want a better packing, the `EnumArray` type will offset by a min.